docs/designs/ps_design.md

```diff
@@ -0,0 +1,422 @@
+
+The interfaces of PServer could be like this:
+
+
+```python
+class PServer(elasticdl_pb2_grpc.PServerServicer):
+    def __init__(self, kvstore, grad_queue, opt):
+        self.kvstore = KVStore()
+        self.grad_queue = GradientQueue()
+        self.opt = Optimizer(opt, self.kvstore, self.grad_queue)
+
+    def push_model(self, request, _):
+        pass
+
+    def push_gradient(self, request, _):
+        pass
+        
+    def pull_variable(self, request, _):
+        pass
+
+    # embedding param is handled lazily
+    def pull_embedding_vector(self, reques, _):
+        pass
+```
+
+### Checkpoint and Serving
+
+Master will send signal to pservers to make checkpoint. Each pserver will save parameters in its current KVStore to a distributed file system.
+
+Since a pserver only has a subset of the whole model, we have to merge these submodels to get final model for serving.
+
+
+## Interactions among Master, PS and Worker
+The following events involve interactions among the master, workers and PS:
+
+* The master starts PS.
+* Initialization of parameters in PS.
+* Relaunch of PS.
+* Workers get model variables from PS.
+* Workers push gradients to PS.
+* PS reports submodel version to the master.
+* The master tells PS to save checkpoint.
+
+### The master starts PS
+When an ElasticDL task starts, `master.main` is responsible for starting PS as a Kubernetes service. Through Kubernetes service, we can fix domain name for every PS node.
+
+After starting PS, `master.main` starts the master servicer and workers, and tells them the domain names of all PS nodes. For PS with embedding replicas, every PS node also needs to know the domain name of its replicas.
+
+### Initialization of parameters in PS
+PS does not have any model parameters and model meta info after starting. Model meta info includes dimension of embedding layers, initialization methods of embedding vectors, initialization methods of slot variables in optimizer.
+
+There are two ways for PS to get model parameters and model meta info, one is to read from a checkpoint file, one is to obtain them from workers.
+
+When `master.main` starts PS, `master.main` decides how to initialize PS according to the arguments specified by users (train from scratch or train from checkpoint).
+
+If `master.main` passes an argument specifying the checkpoint file name to PS, PS reads from the checkpoint. Every PS node scans the variables and embedding vectors in the checkpoint, and saves the part which belong to it after parameter sharding.
+
+Otherwise, PS obtains model variables and model meta info from workers. PS does nothing but waiting for the first `pull_variable` call from worker. In the reponse of `pull_variable` call, PS tells the worker to initialize model, and report model variables and meta info to the PS. This process can be represented in the pseudocode:
+
+```python
+message PullModelRequest{
+    enum MethodType {
+        MINIMUM = 0;
+        FIXED = 1;
+    }
```

Comment:
Not very sure what's the usage of the `MethodType`,  the definition is copied from the current code base.